File: DP/src/main/java/Solution152.java
public class Solution152 {
    public int maxProduct(int[] nums) {
        int n = nums.length;

        int[] optMax = new int[n];
        int[] optMin = new int[n];
        optMax[0] = nums[0];
        optMin[0] = nums[0];

        for (int i = 1; i < n; i++) {
            optMax[i] = Math.max(Math.max(optMax[i - 1] * nums[i], optMin[i - 1] * nums[i]), nums[i]);

            optMin[i] = Math.min(Math.min(optMax[i - 1] * nums[i], optMin[i - 1] * nums[i]), nums[i]);
        }

        int ans = nums[0];

        for (int i = 0; i < n; i++) {
            ans = Math.max(ans, optMax[i]);
        }

        return ans;

    }
}

File: DP/src/main/python/solution152.py
from typing import List

# https://leetcode-cn.com/problems/maximum-product-subarray/


class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n = len(nums)
        opt = [[float("INF"), float("-INF")] for _ in range(n)]

        opt[0] = [nums[0], nums[0]]

        for i in range(1, n):
            opt[i][0] = min(opt[i - 1][0] * nums[i], opt[i - 1][1] * nums[i], nums[i])
            opt[i][1] = max(opt[i - 1][0] * nums[i], opt[i - 1][1] * nums[i], nums[i])

        # print(opt)
        return max(sum(opt, []))
